crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -211,10 +235,7 @@ impl Actor for MqttActor {
                         Ok(None) | Err(RuntimeRequest::Shutdown) => break,
                     }
                 }
-                Some(message) = mqtt_client.received.next() => {
-                    self.messages.send(message).await?
-                },
-                else => break,
```

Comment:
The else arm has been squashed into the task inside `tokio::spawn` actually, as this only affects the outgoing sender. `else` in `tokio::select` is called when all the other branches are disabled (i.e. we've received values for the awaited futures, but we none of them matched the provided patterns). Given the first pattern is just a variable and will therefore match anything, only `Some(message) = mqtt_client.received.next()` can be disabled, and this occurs when calling `.next()` returns `None`, which happens when `mqtt_client.received` is closed.

Without changing what the original code does (but absolutely changing what happens if we were to add a new variant to `RuntimeRequest`, which would be treated like shutdown without any warning), we could change it to the following to make full use of `else`.

```rust
tokio::select! {
    Ok(Some(message)) = self.messages.try_recv() => {
        mqtt_client
            .published
            .send(message)
            .await
            .map_err(Box::new)?
    }
    Some(message) = mqtt_client.received.next() => {
        self.messages.send(message).await?
    },
    else => break,
}
```

In general though, avoiding pattern matching at the `tokio::select` level is a useful thing to do, as like in this example, there are many cases where it work like match when initially written, but then won't work if changes are made to the code surrounding it. Notably, if we had something like 

```rust
tokio::select! {
    RuntimeRequest::Shutdown => self.signals.next(),
    // more cases
}

enum RuntimeRequest {
    Shutdown,

    // Suppose we then add a future variant like
    AbsolutelyDoNotShutdownJustAcknowledgeIExist,
}
```

`tokio::select` will accept this, but when we receive a `RuntimeRequest::AbsolutelyDoNotShutdownJustAcknowledgeIExist`, the `self.signals.next()` branch has failed to match, which just causes that branch to block forever, and we wouldn't receive any shutdown signals. If however we were doing this in a `match` on `self.signals.next().await`, this absolutely wouldn't compile until we handled both `RuntimeRequest` values (this is the argument in favour of keeping the internal match for the first `tokio::select` branch). When you have a bunch of branches and some of them are failing for this reason, it can be really quite tricky to notice if you're not careful.

In general, I think `tokio::select` is quite a lot riskier than thin-edge gives it credit for (it's also very useful, mind you). It looks similar to a `match`, but where `match` generates a compiler error, `tokio::select` just blocks (one possible code path), which is very likely not the desired behaviour.

Reply:
Thanks for that detailed explanation.

> which just causes that branch to block forever

That's definitely concerning, as we have such pattern matching `tokio::select!` matching patterns in a lot more places and I'm not sure if we have the `else` fallback everywhere to avoid such "blocking" issues.

> If however we were doing this in a match on self.signals.next().await, this absolutely wouldn't compile until we handled both RuntimeRequest values

I'm assuming that's the recommended way to do it. Just do `tokio:select!` first and inside the processing arm, `match` for sub-variants, rather than pushing those matches to the top level.

Reply:
> That's definitely concerning, as we have such pattern matching tokio::select! matching patterns in a lot more places and I'm not sure if we have the else fallback everywhere to avoid such "blocking" issues.

`else` will stop you blocking when you cannot feasibly make any more progress, but if there's a branch which can make progress, you can still have issues, and they can be pretty subtle.

> I'm assuming that's the recommended way to do it. Just do `tokio:select!` first and inside the processing arm, `match` for sub-variants, rather than pushing those matches to the top level.

Yeah I think so. I would consider breaking the functionality down into separate async functions/blocks and using [`futures::future::select_all`](https://docs.rs/futures/latest/futures/future/fn.select_all.html), although I think this tends to interact badly with mutable borrows.